### PR TITLE
`hash_newtype` cleanup

### PR DESCRIPTION
**Pull request by Kixunil**
Removed unused macro and improved documentation to address review of #1659 - see commits. I also added a note about recursion.

### CODE REVIEW

**Comment by apoelstra** [comment]
NIce! These are super helpful.

**Review by apoelstra — DISMISSED** [review]
ACK afb4965e14ad39e39c0cc6cc4301de4675ac06c1

**Comment by sanket1729** [comment]
I did not spend time trying to understand the macros or why certain ways fail. I can ACK without understanding the comments under "Does not harm the current code" and "improves documentation from a contributor". 


Will wait a few days if others want to spend time understanding the rationale of comments and carefully ACK it.

**Comment by apoelstra** [comment]
@sanket1729 so, these macros are very new. They were introduced in #1659 and this PR adds documentation at the request of tcharding (who reviewed the code more carefully than I did evidently).

File: hashes/src/util.rs [hunk 1]
```diff
@@ -156,6 +156,37 @@ macro_rules! engine_input_impl(
...
 /// }
 /// ```
+///
+/// Note: the macro is internally recursive. If you use too many attributes (> 256 tokens) you may
+/// hit recursion limit. If you have so many attributes for a good reason, just raising the limit
+/// should be OK.
+///
+// Ever heard of legendary comments warning developers to not touch the code? Yep, here's another
+// one. The following code is written the way it is for some specific reasons. If you think you can
+// simplify it, I suggest spending your time elswhere.
```
Comment (tcharding) [inline hashes/src/util.rs:166]: lolz, I chuckled reading this, love it.

File: hashes/src/util.rs [hunk 2]
```diff
@@ -302,13 +343,14 @@ macro_rules! hash_newtype_struct {
...
-macro_rules! hash_newtype_our_attrs {
-    (hash_newtype($($attr:tt)*)) => { $($attr)* };
-    ($($ignore:tt)*) => {};
-}
-
+// Extracts `hash_newtype(forward)` and `hash_newtype(backward)` attributes if any and turns them
+// into bool, defaulting to `DISPLAY_BACKWARD` of the wrapped type if the attribute is omitted.
+//
+// Once an appropriate attribute is found we pass the remaining ones into another macro to detect
+// suplicates/conflicts and report an error.
```
Comment (tcharding) [inline hashes/src/util.rs:350]:
```suggestion
// duplicates/conflicts and report an error.
```

**Review by tcharding — APPROVED** [review]
One typo, else good to go. Thanks man!

**Comment by Kixunil** [comment]
Noticed and fixed another typo and added a note about quadratic complexity of TT muncher.

**Review by apoelstra — APPROVED** [review]
ACK 8ccfb412c18abaabc4365ab6a2e61adb6a8951f2

**Review by tcharding — APPROVED** [review]
ACK 8ccfb412c18abaabc4365ab6a2e61adb6a8951f2

**Pull request merged**